src/storage/serde.rs

```diff
@@ -1,115 +1,232 @@
-/// `Empty`: The result when caller requested to deserialize and the length of data bytes is zero.
-/// `Item`: The result when caller requested to deserialize.
+/// `Skip`:  The result when caller requested not to deserialize.
+/// `Item`:  The result when caller requested to deserialize. `None` means a tombstone.
+/// 
+/// The `usize` item: Total count of bytes that are read from file.
 enum FileItem {
     EOF,
     Skip(usize),
-    Empty(usize),
-    Item(usize, Vec<u8>),
+    Item(usize, Option<Datum>),
 }
 
-/// From file reads one item, which consists of:
-/// 1) prefix bytes indicating the length of data to follow
-/// 2) bytes taken up by the data
-///
-/// The returned bytes = the prefix bytes + the data bytes
-///
-/// An error is returned iff the exact amount of expected bytes cannot be read from file.
 fn read_item(file: &mut File, deser: bool) -> Result<FileItem> {
-    const PRE_SIZE: usize = size_of::<usize>();
-    let mut buf = [0u8; PRE_SIZE];
-    let read_size = file.read(&mut buf)?;
+    let mut span_size_buf = [0u8; size_of::<usize>()];
+    let mut read_size = file.read(&mut span_size_buf)?;
     if read_size == 0 {
         return Ok(FileItem::EOF);
-    } else if read_size != buf.len() {
-        return Err(anyhow!("File is corrupted."));
+    } else if read_size != span_size_buf.len() {
+        return Err(anyhow!(
+            "Unexpected EOF while reading the beginning of an item."
+        ));
     }
 
-    let datum_size = usize::from_le_bytes(buf);
-    let total_size = PRE_SIZE + datum_size;
+    let span_size = usize::from_le_bytes(span_size_buf);
+    read_size += span_size;
     if deser {
-        if datum_size == 0 {
-            Ok(FileItem::Empty(total_size))
-        } else {
-            let mut buf = vec![0u8; datum_size];
-            file.read_exact(&mut buf)?;
-            Ok(FileItem::Item(total_size, buf))
-        }
+        let mut datum_type_buf = [0u8; size_of::<DatumTypeInt>()];
+        file.read_exact(&mut datum_type_buf)?;
+        let datum_type = DatumTypeInt::from_le_bytes(datum_type_buf);
+
+        let datum_size = span_size - datum_type_buf.len();
+
+        let optdat = deserialize_optdat(datum_type, datum_size, file)?;
+
+        Ok(FileItem::Item(read_size, optdat))
     } else {
-        file.seek(SeekFrom::Current(datum_size as i64))?;
-        Ok(FileItem::Skip(total_size))
+        file.seek(SeekFrom::Current(span_size as i64))?;
+        Ok(FileItem::Skip(read_size))
     }
 }
 
-/// `KV`: The key and value are each an `Option`. Each `Option` reflects whether the caller requested to deserize that part. (The optionality has nothing to do with tombstones.)
+/// `KV`: The key and value are each an `Option`, reflecting whether the caller requested to deserize that part. This optionality is separate from tombstone; tombstone is captured within the `Value` type.
 pub enum FileKeyValue {
     EOF,
     KV(usize, Option<Key>, Option<Value>),
```

Comment:
Let's think about the use cases that we have so far when reading out of a file (whether commit log or sstable):

- initialization
    - read commit log
        deserialize every key and every value; do not need offset
    - read sstable
        sparsely deserialize key; always skip value; need offset
- flushing
    - write sstable
- compacting
    - read sstable
        deserialize every key and every value; do not need offset
    - write sstable
- searching within sstable
    - read sstable
        first seek to an offset in file
        deserialize every key. Then, when a condition is met (key equivalence), deserialize the latest value
        do not need offset

`read_kv()` is a one-size-fits-all method that targets all above use cases. The benefit of having a method like that is that the logic for checking EOF and tombstone is in one place.

I see a couple of solutions:
- Always deserialize key and value. I.e. always use the KeyValueIterator. This would hurt performance but would make code cleaner.
- Create separate read_kv methods for each use case. There are only 3 varieties, and I don't foresee additional use cases. So we just need to split read_kv into 2 varieties.
- In addition to above, separate read_key and read_value.
- In addition to above, remove `KeyValueIterator`, for clarity.